content/advent-2015/building-static-linuxes.md

```diff
@@ -0,0 +1,94 @@
+
+To boot our new kernel and make sure everything works, just use QEMU. You should
+get a `Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(0,0)`
+which makes complete sense, as we haven't added any sort of root FS.
+
+<script type="text/javascript" src="https://asciinema.org/a/12i116g0kltwwxn2c8jwmsq81.js" id="asciicast-12i116g0kltwwxn2c8jwmsq81" async></script>
+
+Using a tool within `libguestfs`, `virt-make-fs` we can create a QEMU compatible
+`.qcow2` from a directory (mine is a directory named `null`) and boot with it.
+
+<script type="text/javascript" src="https://asciinema.org/a/bfk43pq02kdq0hq4das6q70f3.js" id="asciicast-bfk43pq02kdq0hq4das6q70f3" async></script>
+
+Looking good, but we got another kernel panic - `Kernel panic - not syncing: No
+working init found. Try passing init= option to kernel. See Linux
+Documentation/init.txt for guidance.`. Luckily, this one's easy to solve, as,
+funnily enough, any binary can be used as the kernel `init` - Go just so happens
+to make wonderful, static binaries.
+
+# Termboy-go
+
+As such, we need a binary to be our `init`. So let's clone, then build a static
+copy of `termboy-go` with `go build -a --ldflags="-s -X -linkmode external -extldflags -static"`,
+the extra LDFLAGS are needed, as `termboy` uses a bit of CGo to handle keyboard
+input. You'll probably need to slightly tweak termboy-go (hardcode in the ROM
+path you want to use), but this isn't particularly difficult.
+
+Unfortunately, `termboy-go` depends on having a copy of `setfont` in $PATH. This
+is part of how it manages to render pixel-perfect 2D graphics in text-mode consoles.
+Luckily, you can find a copy prebuilt from [minos.io](http://s.minos.io/bifrost/x86_64/),
+or rip apart a copy with `statify` or `ermine` (convert dynamic binaries to static).
+
+Copy this `termboy-go` into the directory you used `virt-make-fs` on, along with
+a Gameboy Color ROM, and the copy of `setfont` you just obtained.
+
+After this, you should be able to just boot your new gameboy-as-a-OS ->
+
+EMBED YOUTUBE VIDEO HERE ONCE IT UPLOADS
+
+# Conclusion
+
+Just about any Go program can be built into a static binary, making it the perfect
+choice for developing embedded systems, like this gameboy-emulator system we built
+today. This single-OS approach may work well for running high-performance applications
+in production, if you're interested in doing so, drop me a line.
+
+However, there are many, far more ridiculous things that can be done using Go to
+tackle early-userland OS problems - I have a bit of a distributed device manager
+laying around, that synchronizes `/dev/` across all devices running it... and that's
+just the beginning of the things that are possible here. Go is _fantastic_ for
+this use case.
+
+Feel free to get in touch with me via email (fortytw2 at gmail) for just about
+any reason, especially if you have questions about this post, or find me @fortytw2
+on twitter/github.
```

Comment:
You might want to consider inserting links to GitHub and Twitter so a reader can just click through to them.